--- wget-1.10/src/log.c.logstdout	2005-05-15 07:15:08.000000000 -0600
+++ wget-1.10/src/log.c	2005-07-04 15:03:39.000000000 -0600
@@ -537,12 +537,19 @@
 {
   if (file)
     {
-      logfp = fopen (file, appendp ? "a" : "w");
-      if (!logfp)
-	{
-	  fprintf (stderr, "%s: %s: %s\n", exec_name, file, strerror (errno));
-	  exit (1);
-	}
+      if (strcmp(file, "-"))
+        {
+          logfp = fopen (file, appendp ? "a" : "w");
+          if (!logfp)
+          {
+	    fprintf (stderr, "%s: %s: %s\n", exec_name, file, strerror (errno));
+	    exit (1);
+	  }
+        }
+      else
+        {
+          logfp = stdout;
+        }
     }
   else
     {
